contracts/token/MEP804.sol

```diff
@@ -0,0 +1,246 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.18;
+
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "./IMEP804.sol";
+import "./IMEP803.sol";
+import "./MEP802.sol";
+import "./IERC6551Account.sol";
+
+/**
+ * @title RewardContract
+ * @author Abiodun Awoyemi
+ * @notice This contract set up reward formula, submit and calculate reward, and user claims reward.
+ */
+contract RewardContract is IMEP804, ERC20, ReentrancyGuard {
+    // STRUCT
+    struct SensorMiningData {
+        uint256 tokenId;
+        address sensorProfileContractAddress;
+        uint256 healthFactor;
+        uint256 miningPower;
+    }
+
+    // STATE VARIABLES
+    address public owner;
+    address public businessOwner;
+    address public applicationContractAddress;
+    ProvisioningContract public provisioningContractAddress;
+    address public lpwanAddress;
+    address[] public sensorProfileAddresses;
+    string public xToEarnFormulaJSON;
+    uint256 public poolAmountOfThisCycle;
+    bool public healthStatus;
+    uint256 public totalMiningPower;
+
+    // MAPPINGS
+    mapping(address => uint256) public reward;
+    mapping(string => uint256) public tierStrength;
+    mapping(string => mapping(address => SensorMiningData)) public miningData;
+
+    /**
+     * @dev Constructor to initialize the contract with essential parameters and mint initial reward tokens.
+     */
+    constructor(
+        address _applicationContractAddress,
+        address _provisioningContractAddress,
+        address _businessOwner,
+        address _lpwanAddress,
+        address[] memory _sensorProfileAddresses,
+        string memory _xToEarnFormulaJSON,
+        string memory _name,
+        string memory _symbol,
+        uint256 _totalRewardAmount
+    ) ERC20(_name, _symbol) {
+        owner = msg.sender;
+        businessOwner = _businessOwner;
+
+        applicationContractAddress = _applicationContractAddress;
+
+        provisioningContractAddress = ProvisioningContract(_provisioningContractAddress);
+
+        lpwanAddress = _lpwanAddress;
+        sensorProfileAddresses = _sensorProfileAddresses;
```

Comment:
Each application can have multiple tiers, therefore there will be multiple sensor profile contract addresses linked to each MEP804. Why would we require one sensor profile here?

Reply:
It is an array of sensor profile addressese